aws/resource_aws_appsync_graphql_api_test.go

```diff
@@ -0,0 +1,112 @@
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAwsAppsyncGraphqlApi_basic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAwsAppsyncGraphqlApiDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAppsyncGraphqlApiConfig_apikey(acctest.RandString(5)),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsAppsyncGraphqlApiExists("aws_appsync_graphql_api.test_apikey"),
+					resource.TestCheckResourceAttrSet("aws_appsync_graphql_api.test_apikey", "arn"),
+				),
+			},
+			{
+				Config: testAccAppsyncGraphqlApiConfig_iam(acctest.RandString(5)),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsAppsyncGraphqlApiExists("aws_appsync_graphql_api.test_iam"),
+					resource.TestCheckResourceAttrSet("aws_appsync_graphql_api.test_iam", "arn"),
+				),
+			},
+			{
+				Config: testAccAppsyncGraphqlApiConfig_cognito(acctest.RandString(5)),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsAppsyncGraphqlApiExists("aws_appsync_graphql_api.test_cognito"),
+					resource.TestCheckResourceAttrSet("aws_appsync_graphql_api.test_cognito", "arn"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAwsAppsyncGraphqlApiDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).appsyncconn
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_appsync_graphql_api" {
+			continue
+		}
+
+		input := &appsync.GetGraphqlApiInput{
+			ApiId: aws.String(rs.Primary.ID),
+		}
+
+		_, err := conn.GetGraphqlApi(input)
+		if err != nil {
+			if isAWSErr(err, appsync.ErrCodeNotFoundException, "") {
+				return nil
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func testAccCheckAwsAppsyncGraphqlApiExists(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("Not found: %s", name)
+		}
+
+		return nil
+	}
+}
+
+func testAccAppsyncGraphqlApiConfig_apikey(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_appsync_graphql_api" "test_apikey" {
+  authentication_type = "API_KEY"
+  name = "tf_appsync_%s"
+}
+`, rName)
+}
+
+func testAccAppsyncGraphqlApiConfig_iam(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_appsync_graphql_api" "test_iam" {
+  authentication_type = "AWS_IAM"
+  name = "tf_appsync_%s"
+}
+`, rName)
+}
+
+func testAccAppsyncGraphqlApiConfig_cognito(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_cognito_user_pool" "test" {
+  name = "tf-%s"
+}
+
+resource "aws_appsync_graphql_api" "test_cognito" {
+  authentication_type = "AMAZON_COGNITO_USER_POOLS"
+  name = "tf_appsync_%s"
+  user_pool_config {
+    aws_region = "us-west-2"
```

Comment:
Although our tests commonly run in `us-west-2` I think we should keep the test region-agnostic. Do you mind adding `aws_region` data source here?

https://www.terraform.io/docs/providers/aws/d/region.html